drake/core/Gradient.h

```diff
@@ -18,55 +18,153 @@ namespace Drake {
   typedef TaylorMatd<Eigen::Dynamic,Eigen::Dynamic,Eigen::Dynamic> TaylorMatXd;
 
-  /** initTaylorVecXd
-   * @brief initializes the vector with x=val and dx=eye(numel(val))
+  /** \brief The appropriate AutoDiffScalar gradient type given the value type and the number of derivatives at compile time
    */
-  template <typename DerivedA, typename DerivedB>
-  void initTaylorVecXd(const Eigen::MatrixBase<DerivedA>& val, Eigen::MatrixBase<DerivedB>& auto_diff_vec) {
-    for (int i=0; i<val.size(); i++) {
-      auto_diff_vec(i) = Eigen::AutoDiffScalar<typename DerivedB::Scalar::DerType>(val(i),val.size(),i);
+  template<typename Derived, int Nq>
+  using AutoDiffMatrixType = Eigen::Matrix<TaylorVard<Nq>, Derived::RowsAtCompileTime, Derived::ColsAtCompileTime, Derived::Options, Derived::MaxRowsAtCompileTime, Derived::MaxColsAtCompileTime>;
+
+
+  /** \brief initializes an autodiff matrix given a matrix of values and gradient matrix
+   * \param[in] val value matrix
+   * \param[in] gradient gradient matrix; the derivatives of val(j) are stored in row j of the gradient matrix.
+   * \param[out] autodiff_matrix matrix of AutoDiffScalars with the same size as
+   */
+  template<typename Derived, typename DerivedGradient, typename DerivedAutoDiff>
+  void initializeAutoDiffGivenGradientMatrix(const Eigen::MatrixBase<Derived> &val, const Eigen::MatrixBase<DerivedGradient> &gradient, Eigen::MatrixBase<DerivedAutoDiff> &auto_diff_matrix)
+  {
+    static_assert(Derived::SizeAtCompileTime == DerivedGradient::RowsAtCompileTime, "gradient has wrong number of rows at compile time");
+    assert(val.size() == gradient.rows() && "gradient has wrong number of rows at runtime");
+    typedef AutoDiffMatrixType<Derived, DerivedGradient::ColsAtCompileTime> ExpectedAutoDiffType;
+    static_assert(ExpectedAutoDiffType::RowsAtCompileTime == DerivedAutoDiff::RowsAtCompileTime, "auto diff matrix has wrong number of rows at compile time");
+    static_assert(ExpectedAutoDiffType::ColsAtCompileTime == DerivedAutoDiff::ColsAtCompileTime, "auto diff matrix has wrong number of columns at compile time");
+    static_assert(std::is_same<typename DerivedAutoDiff::Scalar, typename ExpectedAutoDiffType::Scalar>::value, "wrong auto diff scalar type");
+
+    typedef typename Eigen::MatrixBase<DerivedGradient>::Index Index;
+    auto_diff_matrix.resize(val.rows(), val.cols());
+    auto num_derivs = gradient.cols();
+    for (Index row = 0; row < auto_diff_matrix.size(); row++) {
+      auto_diff_matrix(row).value() = val(row);
+      auto_diff_matrix(row).derivatives().resize(num_derivs, 1);
+      auto_diff_matrix(row).derivatives() = gradient.row(row).transpose();
     }
   }
 
-  template <typename Derived>
-  TaylorVecXd initTaylorVecXd(const Eigen::MatrixBase<Derived>& val) {
-    TaylorVecXd auto_diff_vec(val.rows());
-    initTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  /** \brief initialize a single autodiff matrix given the corresponding value matrix.
```

Comment:
please document very clearly here what happens to the gradient vector.  eye()?  zero()?